RPC: allow to track coins by parent descriptors

darosior opened the pull request:
Wallet descriptors are useful for applications using the Bitcoin Core wallet as a backend for tracking coins, as they allow to track coins for multiple descriptors in a single wallet. However there is no information currently given for such applications to link a coin with an imported descriptor, severely limiting the possibilities for such applications of using multiple descriptors in a single wallet. This PR outputs the matching imported descriptor(s) for a given received coin in `listsinceblock` (and friends).

It comes from a need for an application i'm working on, but i think it's something any software using `bitcoind` to track multiple descriptors in a single wallet would have eventually. For instance i'm thinking about the BDK project. Currently, the way to achieve this is to import raw addresses with labels and to have your application be responsible for wallet things like the gap limit.

I'll add this to the output of `listunspent` too if this gets a few Concept ACKs.

darosior committed:
wallet: allow to fetch the wallet descriptors for a given Script

We currently expose a method to get the signing providers, which allows
to infer a descriptor from the scriptPubKey. But in order to identify
"on" what descriptor a coin was received, we need access to the
descriptors that were imported to the wallet.

darosior committed:
rpc: output wallet descriptors for received entries in listsinceblock

The descriptor wallets allow an application to track coins of multiple
descriptors in a single wallet. However, such an application would not
previously be able to (easily) tell what received coin "belongs" to what
descriptor.

This commit tackles this issues by adding a "wallet_desc" entry to the
entries for received coins in 'listsinceblock'.

w0xlt commented on src/wallet/wallet.cpp:
```diff
@@ -3249,6 +3249,17 @@ std::unique_ptr<SigningProvider> CWallet::GetSolvingProvider(const CScript& scri
+    std::vector<WalletDescriptor> descs;
+    for (const auto spk_man: GetScriptPubKeyMans(script)) {
+        if (const auto desc_spk_man = dynamic_cast<DescriptorScriptPubKeyMan*>(spk_man)) {
+            descs.push_back(desc_spk_man->GetWalletDescriptor());
```
Suggestion to avoid the error mentioned in https://github.com/bitcoin/bitcoin/pull/25504/files#r910264911.

```suggestion
            LOCK(desc_spk_man->cs_desc_man);
            descs.push_back(desc_spk_man->GetWalletDescriptor());
```

w0xlt reviewed (approved):
Approach ACK

darosior commented on src/wallet/wallet.cpp:
```diff
@@ -3249,6 +3249,17 @@ std::unique_ptr<SigningProvider> CWallet::GetSolvingProvider(const CScript& scri
+    std::vector<WalletDescriptor> descs;
+    for (const auto spk_man: GetScriptPubKeyMans(script)) {
+        if (const auto desc_spk_man = dynamic_cast<DescriptorScriptPubKeyMan*>(spk_man)) {
+            descs.push_back(desc_spk_man->GetWalletDescriptor());
```
Thanks

darosior commented on src/wallet/wallet.cpp:
```diff
@@ -3249,6 +3249,17 @@ std::unique_ptr<SigningProvider> CWallet::GetSolvingProvider(const CScript& scri
+{
+    std::vector<WalletDescriptor> descs;
+    for (const auto spk_man: GetScriptPubKeyMans(script)) {
+        if (const auto desc_spk_man = dynamic_cast<DescriptorScriptPubKeyMan*>(spk_man)) {
```
The point here is to filter only descriptor spkmans, which can only be done dynamically.

darosior commented on src/wallet/receive.cpp:
```diff
@@ -265,7 +265,7 @@ void CachedTxGetAmounts(const CWallet& wallet, const CWalletTx& wtx,
         }
 
-        COutputEntry output = {address, txout.nValue, (int)i};
+        COutputEntry output = {address, std::move(txout.scriptPubKey), txout.nValue, (int)i};
```
Oh, yeah, i wanted to revisit this but forgot and pushed.. It's invalid since the caller will use the `scriptPubKey` (for instance in `WalletTxToJSON`).

kristapsk commented:
Concept ACK. JoinMarket currently imports each address into watchonly manually and assigns same label (JM wallet id) to them. So this is needed for JoinMarket to support ranged descriptors in future instead.

achow101 commented:
To be clear, this affects more than just imported descriptors. It outputs the parent descriptor for each scriptPubKey, and so it will as well for the automatically generated descriptors.

achow101 commented on src/wallet/rpc/transactions.cpp:
```diff
@@ -322,6 +322,16 @@ static void MaybePushAddress(UniValue & entry, const CTxDestination &dest)
 }
 
+//! Fetch wallet descriptors matching this scriptPubKey.
+static void MaybePushWalletDescriptors(const CWallet& wallet, const CScript& script_pubkey, UniValue& entry)
```
In 8976cd89051d262bfed2498de5aa9ed190e3b331 "rpc: output wallet descriptors for received entries in listsinceblock"

nit: There's no maybe here, it will always push `wallet_descs`.

achow101 commented on src/wallet/rpc/transactions.cpp:
```diff
@@ -436,7 +447,11 @@ static const std::vector<RPCResult> TransactionDescriptionString()
            {RPCResult::Type::STR, "bip125-replaceable", "(\"yes|no|unknown\") Whether this transaction could be replaced due to BIP125 (replace-by-fee);\n"
-               "may be unknown for unconfirmed transactions not in the mempool."}};
+               "may be unknown for unconfirmed transactions not in the mempool."},
+           {RPCResult::Type::ARR, "wallet_descs", /*optional=*/true, "Only if 'category' is 'received'. List of imported descriptors matching the scriptPubKey of this coin.", {
```
In 8976cd89051d262bfed2498de5aa9ed190e3b331 "rpc: output wallet descriptors for received entries in listsinceblock"

Not just import descriptors

achow101 commented on src/wallet/rpc/transactions.cpp:
```diff
@@ -728,6 +743,9 @@ RPCHelpMan gettransaction()
                                     "'send' category of transactions."},
                                 {RPCResult::Type::BOOL, "abandoned", /*optional=*/true, "'true' if the transaction has been abandoned (inputs are respendable). Only available for the \n"
                                      "'send' category of transactions."},
+                                {RPCResult::Type::ARR, "wallet_descs", /*optional=*/true, "Only if 'category' is 'received'. List of imported descriptors matching the scriptPubKey of this coin.", {
```
In 8976cd89051d262bfed2498de5aa9ed190e3b331 "rpc: output wallet descriptors for received entries in listsinceblock"

Not just imported descriptors.

jb55 commented:
Concept ACK

darosior committed:
rpc: output parent wallet descriptors for coins in listunspent

darosior commented:
Good to see others have an interest in this. :)
I've addressed the review comments and added the same field to the `listunspent` entries.

Note however that there is another barrier to tracking coins by descriptor on the RPC interface, if you want to track multiple descriptors on the same wallet. `listsinceblock` will not add entries for outputs it considers being change, so if you have two descriptors A and B imported on your wallet and you send a coin from A to B it will not be shown as a deposit for B. I think the only reasonable workaround would be to have yet another argument to `listsinceblock` instructing to not treat change outputs differently (like `listunspent` does).
Ideas welcome, but i'd like to tackle this other issue in another PR.

furszy reviewed (commented):
As `ListTransactions` receives the `wtx`, instead of adding a copy of the `scriptPubKey` inside `COutputEntry` to later use it to call `PushWalletDescriptors(wallet, r.script_pubkey, entry)`, what if you just get the output directly?

In other words, drop 9bfd9a96 and change d8b1e37a (line 398) in this way:

```cpp
PushWalletDescriptors(wallet, wtx.tx->vout.at(r.vout).scriptPubKey, entry);
```

darosior commented:
Heh, indeed it's much cleaner. Thanks, done.

darosior committed:
rpc: add an include_change parameter to listsinceblock

It's useful for an external application tracking coins to not be limited
by our change detection. For instance, for a watchonly wallet with two
descriptors a transaction from one to the other would be considered a
change output and not be included in the result (if the address was not
generated by this wallet).

Sjors commented on src/wallet/rpc/util.cpp:
```diff
@@ -123,6 +123,15 @@ std::string LabelFromValue(const UniValue& value)
+    for (const auto& desc: wallet.GetWalletDescriptors(script_pubkey)) {
+        wallet_descs.push_back(desc.descriptor->ToString());
+    }
+    entry.pushKV("wallet_descs", wallet_descs);
```
Maybe call this `parent_descs`? The term `wallet_` is redundant for a wallet RPC.

The `listunspent` RPC already a field called `desc`. This has the limitation that it's not actually looking at the wallet descriptor.

In a followup PR we may want to add `descs` to get descriptor(s) for an individual address, rather than the full range. See #24003.

darosior replied:
Indeed `parent_descs` is a better name, thanks. Done.

darosior committed:
doc: add releases notes for 25504 (listsinceblock updates)

darosior commented:
I added a new commit to make `listsinceblock` optionally list change outputs as well, and added release notes for this PR. Also rebased due to conflict between the new commit and master.

achow101 commented on src/wallet/rpc/transactions.cpp:
```diff
@@ -303,6 +303,16 @@ static void MaybePushAddress(UniValue & entry, const CTxDestination &dest)
 }
 
+//! Fetch parent descriptors of this scriptPubKey.
+static void PushParentDescriptors(const CWallet& wallet, const CScript& script_pubkey, UniValue& entry)
```
In 2a7f812df2ac5c809b4690f9d956564108f705f8 "rpc: output wallet descriptors for received entries in listsinceblock"

To make it easier to read the diffs, perhaps move this to it's final location when you introduce here rather than moving it again in the next commit.

darosior replied:
Done.

achow101 commented on src/wallet/receive.cpp:
```diff
@@ -223,7 +223,8 @@ CAmount CachedTxGetAvailableCredit(const CWallet& wallet, const CWalletTx& wtx,
                   std::list<COutputEntry>& listReceived,
-                  std::list<COutputEntry>& listSent, CAmount& nFee, const isminefilter& filter)
+                  std::list<COutputEntry>& listSent, CAmount& nFee, const isminefilter& filter,
+                  bool report_change)
```
In 812c45af8ed06c985cb5349b11043952e013003d "rpc: add a report_change parameter to listsinceblock"

nit: s/report_change/include_change for consistency with the rpc parameter.

darosior replied:
Done.

darosior commented:
Friendly ping to reviewers, as i'd love to have this in the next release. :)

achow101 commented:
ACK 9e78836cc383357a02412f30f39609d2ea1b3546

instagibbs commented on test/functional/wallet_listsinceblock.py:
```diff
@@ -423,6 +424,18 @@ def test_desc(self):
+        self.generate(self.nodes[2], 1)
+        coins = self.nodes[2].listsinceblock()["transactions"]
+        assert_equal(len([c for c in coins if c["address"] == addr]), 0)
+        coins = self.nodes[2].listsinceblock(include_change=True)["transactions"]
```
would this have two entries since you have the send(to change), and the normal change output as well?

I think the test should be explicit about it if so

darosior replied:
Yes. What would you suggest? Just asserting there are two entries?

instagibbs replied:
Extract both outputs, assert they are change addresses, I suspect is the best.

darosior replied:
Done. This needed a couple re-arrangements since `listsinceblock` lists also the "sent" coins, and we only want coins since the last block.

I couldn't resist a few drive-by cosmetic changes to the tests, including removing the unneeded block generation.

instagibbs commented on doc/release-notes-25504.md:
```diff
@@ -0,0 +1,7 @@
+
+- The `listsinceblock`, `listtransactions` and `gettransaction` output now contain a new
+  `parent_descs` field for every "receive" entry.
+- A new optional parameter was added to the `listsinceblock` command, allowing to also list outputs
```
Just name the argument here?

darosior replied:
Done.

instagibbs commented:
ACK https://github.com/bitcoin/bitcoin/pull/25504/commits/a6b0c1fcc06485ecd320727fa7534a51a20608c1

achow101 commented:
re-ACK a6b0c1fcc06485ecd320727fa7534a51a20608c1

MarcoFalke commented on src/wallet/rpc/transactions.cpp:
```diff
@@ -623,6 +632,7 @@ RPCHelpMan listsinceblock()
     }
 
     bool include_removed = (request.params[3].isNull() || request.params[3].get_bool());
+    bool include_change = (!request.params[4].isNull() && request.params[4].get_bool());
```
I find it ugly to encode default argument values in the code syntax, as opposed to simply name them with a literal. If the default were to change, one would have to change the code syntax as opposed to simply changing a literal.

```cpp
const bool include_change{p.isNull()?false:p.get_bool()}; // should be clearer now and in the future if the default was changed or refactored out

S3RK commented:
> `listsinceblock` will not add entries for outputs it considers being change, so if you have two descriptors A and B imported on your wallet and you send a coin from A to B it will not be shown as a deposit for B.

@darosior I thought I fixed that in #22929. The test you added also specifically generates change address and not receiving address. Could you elaborate on the use-case here since it differs from the test? `listchange` parameter still could be useful, but I'd like to understand when and if deposits are not listed correctly.

darosior commented:
Sure, the use case is using Bitcoin Core to track coins for multiple descriptors, and any descriptor. There exists a number of applications doing that (and in my opinion, there should be more :p).

#22929 was a good fix for users of the wallet as such, but it was limited to a single active descriptor per output type. It's not a viable solutions for applications that want "just an index of transactions per descriptors". Such an application should not care about the wallet internals: whether the descriptor is active or not, the spkman internal or not, the address in the book or not. No, they just want to know the transactions related to a set of descriptor in a given block range. We can provide this information, and that's what this PR is permitting.

Regarding the test, i generated a change address to simulate an address that was externally generated (not in the address book), and not necessarily from a descriptor that was imported as "active" in the watchonly wallet (therefore may be considered internal, ie change).
